Riot/Modules/Settings/SignOut/SignOutFlowPresenter.swift

```diff
@@ -0,0 +1,162 @@
+// 
+// Copyright 2022 New Vector Ltd
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+import Foundation
+
+@objc protocol SignOutFlowPresenterDelegate {
+    /// The presenter is starting an operation that might take while and the UI should indicate this.
+    func signOutFlowPresenterDidStartLoading(_ presenter: SignOutFlowPresenter)
+    /// The presenter has finished an operation and the UI should indicate this if necessary.
+    func signOutFlowPresenterDidStopLoading(_ presenter: SignOutFlowPresenter)
+    /// The presenter encountered an error and has stopped.
+    func signOutFlowPresenter(_ presenter: SignOutFlowPresenter, didFailWith error: Error)
+}
+
+/// This class provides a reusable component to present the sign out flow
+/// for the current session, including the initial prompt, and any follow-up
+/// key-backup setup that is necessary for the user.
+@objcMembers class SignOutFlowPresenter: NSObject {
+    private let session: MXSession
+    private let presentingViewController: UIViewController
+    
+    private var signOutAlertPresenter = SignOutAlertPresenter()
+    
+    weak var delegate: SignOutFlowPresenterDelegate?
+    
+    init(session: MXSession, presentingViewController: UIViewController) {
+        self.session = session
+        self.presentingViewController = presentingViewController
+        
+        super.init()
+        
+        signOutAlertPresenter.delegate = self
+    }
+    
+    /// Starts the flow without a specific source view. On iPad any popups
+    /// will show from the presenting view controller itself.
+    func start() {
+        start(sourceView: presentingViewController.view)
+    }
+    
+    /// Starts the flow, presenting any popups on iPad from the specified view.
+    func start(sourceView: UIView?) {
+        guard let keyBackup = session.crypto?.backup else { return }
+        
+        signOutAlertPresenter.present(for: keyBackup.state,
+                                      areThereKeysToBackup: keyBackup.hasKeysToBackup,
+                                      from: presentingViewController,
+                                      sourceView: sourceView ?? presentingViewController.view,
+                                      animated: true)
```

Comment:
To fix the unable to SignOut issue, I suggest you this kind of change :
```suggestion
        // Tchap: Present the SignOut alert despite the lack of keyBackup (for now).
//        guard let keyBackup = session.crypto?.backup else { return }
        let keyBackup = session.crypto?.backup
        
        signOutAlertPresenter.present(for: keyBackup?.state ?? MXKeyBackupStateDisabled,
                                      areThereKeysToBackup: keyBackup?.hasKeysToBackup ?? false,
                                      from: presentingViewController,
                                      sourceView: sourceView ?? presentingViewController.view,
                                      animated: true)
```

Reply:
Done